cloudpathlib/azure/azblobclient.py

```diff
@@ -200,30 +260,24 @@ def _list_dir(
         if prefix and not prefix.endswith("/"):
             prefix += "/"
 
-        yielded_dirs = set()
-
-        # NOTE: Not recursive may be slower than necessary since it just filters
-        #   the recursive implementation
-        for o in container_client.list_blobs(name_starts_with=prefix):
-            # get directory from this path
-            for parent in PurePosixPath(o.name[len(prefix) :]).parents:
-                # if we haven't surfaced this directory already
-                if parent not in yielded_dirs and str(parent) != ".":
-                    # skip if not recursive and this is beyond our depth
-                    if not recursive and "/" in str(parent):
-                        continue
+        if self._check_hns(cloud_path):
+            file_system_client = self.data_lake_client.get_file_system_client(cloud_path.container)  # type: ignore
+            paths = file_system_client.get_paths(path=cloud_path.blob, recursive=recursive)
 
-                    yield (
-                        self.CloudPath(f"az://{cloud_path.container}/{prefix}{parent}"),
-                        True,  # is a directory
-                    )
-                    yielded_dirs.add(parent)
+            for path in paths:
+                yield self.CloudPath(f"az://{cloud_path.container}/{path.name}"), path.is_directory
 
-            # skip file if not recursive and this is beyond our depth
-            if not recursive and "/" in o.name[len(prefix) :]:
-                continue
+        else:
+            if not recursive:
+                blobs = container_client.walk_blobs(name_starts_with=prefix)
+            else:
+                blobs = container_client.list_blobs(name_starts_with=prefix)
 
-            yield (self.CloudPath(f"az://{cloud_path.container}/{o.name}"), False)  # is a file
+            for blob in blobs:
+                # walk_blobs returns folders with a trailing slash
+                blob_path = blob.name.rstrip("/")
+                blob_cloud_path = self.CloudPath(f"az://{cloud_path.container}/{blob_path}")
+                yield blob_cloud_path, isinstance(blob, BlobPrefix)
```

Comment:
Can you rely on `isinstance(blob, BlobPrefix)` to determine whether the blob is a directory even in the recursive case (`list_blobs()`)?

Reply:
Short answer is yes (there are no folders with `list_blobs` if we are not on an HNS storage account, and the HNS accounts are handled above by the ADLS API). I updated to make more explicit.